### PR TITLE
Speed up startup: faster model/abundance parsing, single-reader input files, and buffered linestat.out

### PR DESCRIPTION
**Pull request by lukeshingles**
Follow-up to #492: a set of startup speedups, benchmarked on a real 3D kilonova setup (125000 cells x 2473 nuclide columns: 2.1 GB `model.txt`, 1.5 GB `transitiondata.txt` with 48.8 million lines, 126 MB `abundances.txt`). **Total startup through the end of input reading on that setup drops from 151 s to 35 s (4.3x)** on an Apple M4 Pro with 4 MPI ranks. All outputs remain bit-identical, so no checksum regeneration is needed.

## model.txt and abundances.txt parsing

- `parse_next_token` moves from `input.cc`'s anonymous namespace to `input.h`, keeping the stream-extraction-compatible semantics established in #492 (leading `+` accepted unless followed by another sign, underflow reads as zero, overflow and `nan`/`inf` spellings rejected).
- The per-cell parsing of `model.txt` (geometry columns and the abundance columns in `read_model_radioabundances`) and `abundances.txt` threads a `std::string_view` remainder through `parse_next_token` instead of refilling a `std::istringstream` per cell and extracting through the locale-aware `operator>>`.
- Real-model measurement: `model.txt` + `Ye.txt` read 25 s -> 6 s; `abundances.txt` 9 s -> 7 s.

## One reader per file instead of every rank

Several readers had every MPI rank parse the same file and either discard the result or write identical values into node-shared memory. Now:

- **`model.txt` cell data — global rank 0 only.** Rank 0 fills its node's shared arrays (`modelgrid_input`, nuclide mass fractions), which are broadcast to the other node leaders over `mpi_comm_internode`; rank-local results (first cell id, vmax, 1D velocity grid) are broadcast to all ranks. One filesystem reader total, with the parsed binary data crossing the interconnect instead of every node re-reading a much larger text file.
- **`abundances.txt`, `Ye.txt`, phixs files, `autoion.txt` — node leaders only**, with the rank-local scalars (`NPHIXSPOINTS`, `NPHIXSNUINCREMENT`, `nbfcontinua`, `nbfcontinua_ground`, autoionisation transition count) broadcast within each node.
- **`setup_phixs_list`**: the node-shared continuum list, ion ground-continuum indices, closest-ground-level continua, and bf-estimator index arrays are now filled by node leaders only instead of every rank writing identical values.
- Left as-is: small files parsed into rank-local data that every rank genuinely needs (decay data, collision/Auger data, gamma spectra, vpkt config, `compositiondata.txt`, `input.txt`).

On clusters this also removes `nranks - 1` (model.txt) or `node_nprocs - 1` (per node, other files) concurrent reads of the same file from the shared filesystem.

## Transitions-to-linelist connection

Each level's down and up transition target lists are sorted in ascending order (they are filled from the `(lower, upper)`-sorted transition table), so the per-line lookups in `create_shared_alltranslist` binary search the block instead of linearly scanning it: O(sum of block sizes squared) -> O(nlines log blocksize). Real-model measurement: 6.3 s -> 2.5 s; the advantage grows linearly with block size. The per-line asserts still verify that every line finds its exact entry.

## linestat.out writing

Profiling showed the largest remaining startup phase was `initialise_linestat_file()`: five passes over the linelist with one `std::print` call per value (240+ million calls on the real dataset), with ~75% of samples inside libc++'s per-call is-this-a-terminal check. The output is now formatted into a string buffer and written in 4 MB chunks: **98 s -> 8 s** writing the full 1.3 GB file, byte-identical (linestat.out is required by the artistools plotting scripts and is always written on new runs, as before).

## Initial temperature assignment

Every rank was computing the decayed energy release and initial temperature of every non-empty cell, with only the node leader storing the results. The cells are now striped across the ranks of each node, each rank writing its own disjoint subset of the node-shared temperature arrays, with the diagnostic counts combined by node-communicator reductions. Real-model measurement (4 ranks): **35 s -> 8 s**, scaling with the number of ranks per node.

## Real-dataset benchmark summary (Apple M4 Pro, 4 MPI ranks, warm cache, two runs per binary)

| startup phase | before (merged #492) | after |
| --- | --- | --- |
| adata + transitiondata parse | 6 s | 6 s (optimised in #492) |
| transitions-to-linelist connection | 6.3 s | 2.5 s |
| `model.txt` + `Ye.txt` read | 25 s | 6 s |
| `linestat.out` write (1.3 GB) | 99 s | 8 s |
| `abundances.txt` read | 9 s | 7 s |
| **total to end of input reading** | **151 s** | **35 s** |

## Verification

Same-machine baseline-vs-changed checksum comparisons (`REPRODUCIBLE=ON FASTMATH=OFF MAX_NODE_SIZE=2`), baseline = merged #492, re-run after each change including the final state of the branch:

- kilonova_1d full sequence (newrun + resume + exspec): 1D spherical model, one-line-per-cell format, `abundances.txt`, and `linestat.out` — all output files checksum-identical.
- classicmode_3d first job: 3D Cartesian model in the two-line-per-cell format with the v1 `phixsdata.txt` and VPKT enabled — all output and checkpoint files checksum-identical.

Both runs use 4 MPI ranks with `MAX_NODE_SIZE=2` (two nodes of two ranks), so the internode model broadcast, the intra-node broadcasts, and the non-leader skip paths are all exercised. The real 3D setup also ran end-to-end through timestep 0 packet propagation as an integration check. **No checksum regeneration is needed.**

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: input.h [hunk 1]
```diff
@@ -41,4 +49,52 @@ inline auto get_noncommentline(std::istream& input, std::string& line) -> bool {
...
+[[nodiscard]] inline auto parse_next_token(std::string_view& remainder, T& value) -> bool {
+  constexpr std::string_view whitespace = " \t\r";
+  const auto tokenstart = remainder.find_first_not_of(whitespace);
+  if (tokenstart == std::string_view::npos) {
+    return false;
+  }
+  const auto tokenend = std::min(remainder.find_first_of(whitespace, tokenstart), remainder.size());
+  auto token = remainder.substr(tokenstart, tokenend - tokenstart);
+  if (token.front() == '+') {  // stream extraction accepted a leading plus sign, but from_chars does not
+    token.remove_prefix(1);
```
[the remark this answers is not in the document]
Comment (lukeshingles) [reply input.h]: Fixed in 0d798eef: the leading plus is now only removed when not followed by another sign, so malformed tokens like +-0, +-1e-8, and ++1 are rejected as stream extraction did. Covered by new edge-case tests (21 cases total).

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

**Comment by lukeshingles** [comment]
@codex review

File: grid.cc [hunk 2]
```diff
@@ -2021,10 +2017,11 @@ void read_ejecta_model() {
     }
 
     if (mgi == 0) {
-      first_cellindex = cellnumberin;
-      printlnlog("first_cellindex {}", first_cellindex);
+      first_input_cellid = cellnumberin;
+      printlnlog("first_input_cellid {}", first_input_cellid);
+      assert_always(first_input_cellid == 0 || first_input_cellid == 1);
```
[the remark this answers is not in the document]
Comment (lukeshingles) [reply grid.cc]: Keeping this check: the 0-or-1 restriction on the first input cell id was added deliberately by the maintainer in 3ef2b1d2. Like the level-numbering autodetection in #492, the first cell id is treated as a numbering-base convention (0- or 1-indexed) rather than an arbitrary offset, and an unexpected starting id like 101 now fails fast instead of being silently renumbered.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_
